src/secondary.rs

```diff
@@ -759,6 +788,391 @@ impl<'a, K: Key, V: 'a + Copy> Extend<(K, &'a V)> for SecondaryMap<K, V> {
+    ///
+    /// if let Entry::Occupied(mut o) = sec.entry(k).unwrap() {
+    ///     let v = o.remove();
+    ///     assert_eq!(v, 10);
+    ///     assert_eq!(sec.get(k), None);
+    /// }
+    /// ```
+    pub fn remove(self) -> V {
+        let keyd = self.key.into();
+        let slot = unsafe { self.map.slots.get_unchecked_mut(keyd.idx as usize) };
+        slot.version -= 1;
+        self.map.num_elems -= 1;
+        match std::mem::replace(&mut slot.value, None) {
+            Some(x) => x,
+            None => unsafe { unreachable_unchecked() },
+        }
+    }
+}
+
+/// A view into a vacant entry in a `SecondaryMap`.
+/// It is part of the `Entry` enum.
+#[derive(Debug)]
+pub struct VacantEntry<'a, K: Key, V> {
+    map: &'a mut SecondaryMap<K, V>,
+    key: K,
+}
+
+impl<'a, K: Key, V> VacantEntry<'a, K, V> {
+    /// Gets a reference to the key that would be used when inserting a value
+    /// through the `VacantEntry`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use slotmap::*;
+    /// # use slotmap::secondary::Entry;
+    ///
+    /// let mut sm = SlotMap::new();
+    /// let mut sec: SecondaryMap<_, ()> = SecondaryMap::new();
+    ///
+    /// let k = sm.insert(1);
+    ///
+    /// if let Entry::Vacant(v) = sec.entry(k).unwrap() {
+    ///     assert_eq!(v.key(), &k);
+    /// }
+    /// ```
+    pub fn key(&self) -> &K {
+        &self.key
+    }
+
+    /// Take ownership of the key.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use slotmap::*;
+    /// # use slotmap::secondary::Entry;
+    ///
+    /// let mut sm = SlotMap::new();
+    /// let mut sec: SecondaryMap<_, ()> = SecondaryMap::new();
+    ///
+    /// let k = sm.insert(1);
+    ///
+    /// if let Entry::Vacant(v) = sec.entry(k).unwrap() {
+    ///     assert_eq!(v.into_key(), k);
+    /// }
+    /// ```
+    pub fn into_key(self) -> K {
+        self.key
+    }
+
+    /// Sets the value of the entry with the VacantEntry's key,
+    /// and returns a mutable reference to it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use slotmap::*;
+    /// # use slotmap::secondary::Entry;
+    ///
+    /// let mut sm = SlotMap::new();
+    /// let mut sec = SecondaryMap::new();
+    ///
+    /// let k = sm.insert(1);
+    ///
+    /// if let Entry::Vacant(v) = sec.entry(k).unwrap() {
+    ///     let new_val = v.insert(3);
+    ///     assert_eq!(new_val, &mut 3);
+    /// }
+    /// ```
+    pub fn insert(self, value: V) -> &'a mut V {
+        let key = self.key.into();
+        self.map
+            .slots
+            .extend((self.map.slots.len()..=key.idx as usize).map(|_| Slot {
+                version: 0,
+                value: None,
+            }));
+
+        let slot = &mut self.map.slots[key.idx as usize];
+        self.map.num_elems += 1;
+
+        *slot = Slot {
+            version: key.version.get(),
+            value: Some(value),
+        };
+
+        match slot.value.as_mut() {
```

Comment:
We literally just assigned to `slot` so I would expect simply `slot.value.as_mut().unwrap()` to be optimized to this, avoiding unnecessary `unsafe`.

Reply:
after checking on goldbolt:
https://godbolt.org/z/LFLvyj
Looks like you're right :-)